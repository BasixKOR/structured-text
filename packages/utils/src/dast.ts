export type Node = BlockNode | InlineNode;

export type BlockNode =
  | Root
  | Paragraph
  | Heading
  | Block
  | List
  | ListItem
  | Blockquote
  | Code;

export type InlineNode = Span | Link | ItemLink | InlineItem;

export const rootType = 'root' as const;
export type RootType = 'root';

export type Root = {
  type: RootType;
  children: Array<Paragraph | Heading | List | Code | Blockquote | Block>;
};

export const paragraphType = 'paragraph' as const;
export type ParagraphType = 'paragraph';

export type Paragraph = {
  type: ParagraphType;
  children: Array<InlineNode>;
};

export const headingType = 'heading' as const;
export type HeadingType = 'heading';

export type Heading = {
  type: HeadingType;
  level: 1 | 2 | 3 | 4 | 5 | 6;
  children: Array<InlineNode>;
};

export const listType = 'list' as const;
export type ListType = 'list';

export type List = {
  type: ListType;
  style: 'bulleted' | 'numbered';
  children: Array<ListItem>;
};

export const listItemType = 'listItem' as const;
export type ListItemType = 'listItem';

export type ListItem = {
  type: ListItemType;
  children: Array<Paragraph | List>;
};

export const codeType = 'code' as const;
export type CodeType = 'code';

export type Code = {
  type: CodeType;
  language?: string;
  children: Array<Span>;
};

export const blockquoteType = 'blockquote' as const;
export type BlockquoteType = 'blockquote';

export type Blockquote = {
  type: BlockquoteType;
  children: Array<Paragraph>;
};

export const blockType = 'block' as const;
export type BlockType = 'block';

export type Block = {
  type: BlockType;
  item: string;
};

export const spanType = 'span' as const;
export type SpanType = 'span';

export type Mark =
  | 'strong'
  | 'code'
  | 'emphasis'
  | 'underline'
  | 'strikethrough'
  | 'highlight';

export type Span = {
  type: SpanType;
  marks?: Mark[];
  value: string;
};

export const linkType = 'link' as const;
export type LinkType = 'link';

export type Link = {
  type: LinkType;
  url: string;
  children: Array<Span>;
};

export const itemLinkType = 'itemLink' as const;
export type ItemLinkType = 'itemLink';

export type ItemLink = {
  type: ItemLinkType;
  item: string;
  children: Array<Span>;
};

export const inlineItemType = 'inlineItem' as const;
export type InlineItemType = 'inlineItem';

export type InlineItem = {
  type: InlineItemType;
  item: string;
};

export type WithChildrenNode = Exclude<Node, Span | Block | InlineItem>;

export type DastDocument = {
  schema: 'dast';
  document: Root;
};

export function isHeading(node: Node): node is Heading {
  return node.type === headingType;
}

export function isSpan(node: Node): node is Span {
  return node.type === spanType;
}

export function hasChildren(node: Node): node is WithChildrenNode {
  return 'children' in node;
}
