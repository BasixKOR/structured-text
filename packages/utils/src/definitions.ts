import { Mark, NodeType } from './types';

export const blockquoteNodeType = 'blockquote' as const;
export const blockNodeType = 'block' as const;
export const codeNodeType = 'code' as const;
export const headingNodeType = 'heading' as const;
export const inlineItemNodeType = 'inlineItem' as const;
export const itemLinkNodeType = 'itemLink' as const;
export const linkNodeType = 'link' as const;
export const listItemNodeType = 'listItem' as const;
export const listNodeType = 'list' as const;
export const paragraphNodeType = 'paragraph' as const;
export const rootNodeType = 'root' as const;
export const spanNodeType = 'span' as const;

export const allowedNodeTypes = [
  blockquoteNodeType,
  blockNodeType,
  codeNodeType,
  headingNodeType,
  inlineItemNodeType,
  itemLinkNodeType,
  linkNodeType,
  listItemNodeType,
  listNodeType,
  paragraphNodeType,
  rootNodeType,
  spanNodeType,
];

type AllowedChildren = Record<
  NodeType,
  'text' | 'inlineNodesAndText' | NodeType[]
>;

export const allowedChildren: AllowedChildren = {
  [blockquoteNodeType]: [paragraphNodeType],
  [blockNodeType]: [],
  [codeNodeType]: 'text',
  [headingNodeType]: 'inlineNodesAndText',
  [inlineItemNodeType]: [],
  [itemLinkNodeType]: 'inlineNodesAndText',
  [linkNodeType]: 'inlineNodesAndText',
  [listItemNodeType]: [paragraphNodeType, listNodeType],
  [listNodeType]: [listItemNodeType],
  [paragraphNodeType]: 'inlineNodesAndText',
  [rootNodeType]: [
    blockquoteNodeType,
    codeNodeType,
    listNodeType,
    paragraphNodeType,
    headingNodeType,
  ],
  [spanNodeType]: [],
};

type AllowedAttributes = Record<NodeType, string[]>;

export const allowedAttributes: AllowedAttributes = {
  [blockquoteNodeType]: ['children'],
  [blockNodeType]: ['item'],
  [codeNodeType]: ['language', 'children'],
  [headingNodeType]: ['level', 'children'],
  [inlineItemNodeType]: ['item'],
  [itemLinkNodeType]: ['item', 'children'],
  [linkNodeType]: ['url', 'children'],
  [listItemNodeType]: ['children'],
  [listNodeType]: ['style', 'children'],
  [paragraphNodeType]: ['children'],
  [rootNodeType]: ['children'],
  [spanNodeType]: ['value', 'marks'],
};

export const allowedMarks: Mark[] = [
  'strong',
  'code',
  'emphasis',
  'underline',
  'strikethrough',
  'highlight',
];
