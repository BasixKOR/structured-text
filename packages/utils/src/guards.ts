import {
  allowedNodeTypes,
  blockNodeType,
  blockquoteNodeType,
  codeNodeType,
  headingNodeType,
  inlineBlockNodeType,
  inlineItemNodeType,
  inlineNodeTypes,
  itemLinkNodeType,
  linkNodeType,
  listItemNodeType,
  listNodeType,
  paragraphNodeType,
  rootNodeType,
  spanNodeType,
  thematicBreakNodeType,
} from './definitions';
import {
  Block,
  BlockId,
  Blockquote,
  CdaStructuredTextValue,
  Code,
  Document,
  Heading,
  InlineBlock,
  InlineItem,
  InlineNode,
  ItemLink,
  Link,
  List,
  ListItem,
  Node,
  NodeType,
  Paragraph,
  Record as DatoCmsRecord,
  Root,
  Span,
  StructuredText,
  ThematicBreak,
  WithChildrenNode,
} from './types';

export function hasChildren<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is WithChildrenNode<BlockItemType> {
  return 'children' in node;
}

export function isInlineNode<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is InlineNode<BlockItemType> {
  return (inlineNodeTypes as NodeType[]).includes(node.type);
}

export function isHeading<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Heading<BlockItemType> {
  return node.type === headingNodeType;
}

export function isSpan<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Span {
  return node.type === spanNodeType;
}

export function isRoot<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Root<BlockItemType> {
  return node.type === rootNodeType;
}

export function isParagraph<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Paragraph<BlockItemType> {
  return node.type === paragraphNodeType;
}

export function isList<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is List<BlockItemType> {
  return node.type === listNodeType;
}

export function isListItem<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is ListItem<BlockItemType> {
  return node.type === listItemNodeType;
}

export function isBlockquote<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Blockquote<BlockItemType> {
  return node.type === blockquoteNodeType;
}

export function isBlock<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Block<BlockItemType> {
  return node.type === blockNodeType;
}

export function isInlineBlock<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is InlineBlock<BlockItemType> {
  return node.type === inlineBlockNodeType;
}

export function isCode<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Code {
  return node.type === codeNodeType;
}

export function isLink<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is Link {
  return node.type === linkNodeType;
}

export function isItemLink<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is ItemLink {
  return node.type === itemLinkNodeType;
}

export function isInlineItem<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is InlineItem {
  return node.type === inlineItemNodeType;
}

export function isThematicBreak<BlockItemType = BlockId>(
  node: Node<BlockItemType>,
): node is ThematicBreak {
  return node.type === thematicBreakNodeType;
}

function isObject(obj: unknown): obj is Record<string, unknown> {
  return Boolean(typeof obj === 'object' && obj);
}

export function isNodeType(value: string): value is NodeType {
  return allowedNodeTypes.includes(value as NodeType);
}

export function isNode<BlockItemType = BlockId>(
  obj: unknown,
): obj is Node<BlockItemType> {
  return Boolean(
    isObject(obj) &&
      'type' in obj &&
      typeof obj.type === 'string' &&
      isNodeType(obj.type),
  );
}

export function isCdaStructuredTextValue<
  BlockRecord extends DatoCmsRecord,
  LinkRecord extends DatoCmsRecord,
  InlineBlockRecord extends DatoCmsRecord
>(
  obj: unknown,
): obj is CdaStructuredTextValue<BlockRecord, LinkRecord, InlineBlockRecord> {
  return Boolean(isObject(obj) && 'value' in obj && isDocument(obj.value));
}

/**
 * @deprecated Use isCdaStructuredTextValue instead
 */
export function isStructuredText<
  BlockRecord extends DatoCmsRecord,
  LinkRecord extends DatoCmsRecord,
  InlineBlockRecord extends DatoCmsRecord
>(
  obj: unknown,
): obj is StructuredText<BlockRecord, LinkRecord, InlineBlockRecord> {
  return isCdaStructuredTextValue(obj);
}

export function isDocument<BlockItemType = BlockId>(
  obj: unknown,
): obj is Document<BlockItemType> {
  return Boolean(
    isObject(obj) &&
      'schema' in obj &&
      'document' in obj &&
      obj.schema === 'dast',
  );
}

export function isEmptyDocument(obj: unknown): boolean {
  if (!obj) {
    return true;
  }

  const document =
    isStructuredText(obj) && isDocument(obj.value)
      ? obj.value
      : isDocument(obj)
      ? obj
      : null;

  if (!document) {
    throw new Error(
      'Passed object is neither null, a Structured Text value or a DAST document',
    );
  }

  return (
    document.document.children.length === 1 &&
    document.document.children[0].type === 'paragraph' &&
    document.document.children[0].children.length === 1 &&
    document.document.children[0].children[0].type === 'span' &&
    document.document.children[0].children[0].value === ''
  );
}
