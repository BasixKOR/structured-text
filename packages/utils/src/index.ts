export * from './dast';

export function foo(): string {
  return 'foo';
}
