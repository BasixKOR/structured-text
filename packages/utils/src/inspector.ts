import {
  hasChildren,
  isSpan,
  isHeading,
  isList,
  isCode,
  isLink,
  isItemLink,
  isBlock,
  isInlineBlock,
  isParagraph,
  isDocument,
} from './guards';
import { Node, Document } from './types';

type StructuredTextDocumentOrNode<T> = T | Document;

export interface InspectOptions {
  /**
   * Custom formatter for block and inlineBlock nodes.
   * Receives the item value (either ID string or full object) and returns a string.
   * The returned string can be multi-line - the inspector will handle proper indentation.
   */
  blockFormatter?: (item: unknown) => string;
}

function extractNode<T>(input: StructuredTextDocumentOrNode<T>): T {
  return isDocument(input) ? ((input.document as unknown) as T) : (input as T);
}

export function inspect<T>(
  input: StructuredTextDocumentOrNode<T>,
  options: InspectOptions = {},
): string {
  const rootNode = extractNode(input);
  return buildTree((rootNode as unknown) as Node, '', true, options);
}

function buildTree(
  node: Node,
  prefix: string,
  isLast: boolean,
  options: InspectOptions,
): string {
  const connector = isLast ? '└' : '├';
  const nodeLabel = buildNodeLabel(node, options);

  // Handle multi-line content
  const lines = nodeLabel.split('\n');
  const firstLine = `${prefix}${connector} ${lines[0]}\n`;

  let multiLineContent = '';
  if (lines.length > 1) {
    const contentPrefix = prefix + (isLast ? '  ' : '│ ');
    for (let i = 1; i < lines.length; i++) {
      multiLineContent += `${contentPrefix}${lines[i]}\n`;
    }
  }

  if (!hasChildren(node)) {
    return firstLine + multiLineContent;
  }

  const children = node.children;
  const childPrefix = prefix + (isLast ? '  ' : '│ ');

  let result = firstLine + multiLineContent;
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    const isLastChild = i === children.length - 1;
    result += buildTree(child as Node, childPrefix, isLastChild, options);
  }

  return result;
}

function buildNodeLabel(node: Node, options: InspectOptions): string {
  const metaInfo: string[] = [];
  let content = '';

  if (isSpan(node)) {
    if (node.marks && node.marks.length > 0) {
      metaInfo.push(`marks: ${node.marks.join(', ')}`);
    }
    content = ` "${truncateText(node.value, 50)}"`;
  } else if (isCode(node)) {
    if (node.language) {
      metaInfo.push(`language: "${node.language}"`);
    }
    content = ` "${truncateText(node.code, 50)}"`;
  } else if (isHeading(node)) {
    metaInfo.push(`level: ${node.level}`);
  } else if (isList(node)) {
    metaInfo.push(`style: ${node.style}`);
  } else if (isLink(node)) {
    metaInfo.push(`url: "${node.url}"`);
    if (node.meta && node.meta.length > 0) {
      const metaEntries = node.meta
        .map((m) => `${m.id}="${m.value}"`)
        .join(', ');
      metaInfo.push(`meta: {${metaEntries}}`);
    }
  } else if (isItemLink(node)) {
    metaInfo.push(`item: "${node.item}"`);
    if (node.meta && node.meta.length > 0) {
      const metaEntries = node.meta
        .map((m) => `${m.id}="${m.value}"`)
        .join(', ');
      metaInfo.push(`meta: {${metaEntries}}`);
    }
  } else if (isBlock(node)) {
    if (options.blockFormatter) {
      const formattedContent = options.blockFormatter(node.item);
      content = ` ${formattedContent}`;
    } else {
      metaInfo.push(`item: "${node.item}"`);
    }
  } else if (isInlineBlock(node)) {
    if (options.blockFormatter) {
      const formattedContent = options.blockFormatter(node.item);
      content = ` ${formattedContent}`;
    } else {
      metaInfo.push(`item: "${node.item}"`);
    }
  } else if (isParagraph(node)) {
    if (node.style) {
      metaInfo.push(`style: "${node.style}"`);
    }
  }

  const metaString = metaInfo.length > 0 ? ` (${metaInfo.join(', ')})` : '';
  return `${node.type}${metaString}${content}`;
}

function truncateText(text: string, maxLength: number): string {
  if (text.length <= maxLength) {
    return text;
  }
  return text.substring(0, maxLength - 3) + '...';
}
