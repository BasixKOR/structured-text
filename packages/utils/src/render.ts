import { Node, Record, StructuredText } from './types';
import { hasChildren, isStructuredText } from './guards';
import flatten from 'lodash.flatten';

export class RenderError extends Error {
  node: Node;

  constructor(message: string, node: Node) {
    super(message);
    this.node = node;
    Object.setPrototypeOf(this, RenderError.prototype);
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type TrasformFn = (...args: any[]) => any;

export type RenderResult<
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn
> = ReturnType<H> | ReturnType<T> | ReturnType<F> | null | undefined;

export type RenderContext<
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn,
  N extends Node
> = {
  adapter: Adapter<H, T, F>;
  node: N;
  ancestors: Node[];
  key: string;
  children: Exclude<RenderResult<H, T, F>, null | undefined>[] | undefined;
};

export interface RenderRule<
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn
> {
  appliable: (node: Node) => boolean;
  apply: (ctx: RenderContext<H, T, F, Node>) => RenderResult<H, T, F>;
}

export const renderRule = <
  N extends Node,
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn
>(
  guard: (node: Node) => node is N,
  transform: (ctx: RenderContext<H, T, F, N>) => RenderResult<H, T, F>,
): RenderRule<H, T, F> => ({
  appliable: guard,
  apply: (ctx: RenderContext<H, T, F, Node>) =>
    transform(ctx as RenderContext<H, T, F, N>),
});

export function transformNode<
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn
>(
  adapter: Adapter<H, T, F>,
  node: Node,
  key: string,
  ancestors: Node[],
  renderRules: RenderRule<H, T, F>[],
): RenderResult<H, T, F> {
  const children = hasChildren(node)
    ? (flatten(
        (node.children as Node[])
          .map((node, index) =>
            transformNode(
              adapter,
              node,
              `t-${index}`,
              [...ancestors, node],
              renderRules,
            ),
          )
          .filter((x) => !!x),
      ) as Exclude<RenderResult<H, T, F>, null | undefined>[])
    : undefined;

  const matchingTransform = renderRules.find((transform) =>
    transform.appliable(node),
  );

  if (matchingTransform) {
    return matchingTransform.apply({ adapter, node, children, key, ancestors });
  } else {
    throw new RenderError(
      `Don't know how to render a node with type "${node.type}". Please specify a custom renderRule for it!`,
      node,
    );
  }
}

export type Adapter<
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn
> = {
  renderNode: H;
  renderText: T;
  renderFragment: F;
};

export function render<
  H extends TrasformFn,
  T extends TrasformFn,
  F extends TrasformFn,
  R extends Record
>(
  adapter: Adapter<H, T, F>,
  structuredTextOrNode: StructuredText<R> | Node | null | undefined,
  renderRules: RenderRule<H, T, F>[],
): RenderResult<H, T, F> {
  if (!structuredTextOrNode) {
    return null;
  }

  const result = transformNode(
    adapter,
    isStructuredText<R>(structuredTextOrNode)
      ? structuredTextOrNode.value.document
      : structuredTextOrNode,

    't-0',
    [],
    renderRules,
  );

  return result;
}
