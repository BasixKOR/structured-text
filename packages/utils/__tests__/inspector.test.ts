import { inspect } from '../src/inspector';
import type { Document, Root } from '../src/types';

describe('inspector', () => {
  describe('inspect', () => {
    it('should render a simple root node', () => {
      const simpleRoot: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            children: [{ type: 'span', value: 'Hello World' }],
          },
        ],
      };

      const result = inspect(simpleRoot);
      expect(result).toMatchSnapshot();
    });

    it('should render a document wrapper', () => {
      const document: Document = {
        schema: 'dast',
        document: {
          type: 'root',
          children: [
            {
              type: 'paragraph',
              children: [{ type: 'span', value: 'Test' }],
            },
          ],
        },
      };

      const result = inspect(document);
      expect(result).toMatchSnapshot();
    });

    it('should render multiple children with correct tree structure', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            children: [{ type: 'span', value: 'First paragraph' }],
          },
          {
            type: 'paragraph',
            children: [{ type: 'span', value: 'Second paragraph' }],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render span nodes with marks', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            children: [
              { type: 'span', value: 'Normal text' },
              { type: 'span', marks: ['strong'], value: 'Bold text' },
              {
                type: 'span',
                marks: ['strong', 'emphasis'],
                value: 'Bold italic',
              },
            ],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render heading nodes with level', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'heading',
            level: 1,
            children: [{ type: 'span', value: 'Main Title' }],
          },
          {
            type: 'heading',
            level: 2,
            children: [{ type: 'span', value: 'Subtitle' }],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render list nodes with style', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'list',
            style: 'bulleted',
            children: [
              {
                type: 'listItem',
                children: [
                  {
                    type: 'paragraph',
                    children: [{ type: 'span', value: 'First item' }],
                  },
                ],
              },
              {
                type: 'listItem',
                children: [
                  {
                    type: 'paragraph',
                    children: [{ type: 'span', value: 'Second item' }],
                  },
                ],
              },
            ],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render code nodes with language and content', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'code',
            language: 'javascript',
            code: 'console.log("Hello, World!");',
          },
          {
            type: 'code',
            code: 'echo "No language specified"',
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render link nodes with URL and meta', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            children: [
              { type: 'span', value: 'Visit ' },
              {
                type: 'link',
                url: 'https://example.com',
                children: [{ type: 'span', value: 'this link' }],
              },
              {
                type: 'link',
                url: 'https://meta.example.com',
                meta: [
                  { id: 'title', value: 'Example Site' },
                  { id: 'target', value: '_blank' },
                ],
                children: [{ type: 'span', value: 'link with meta' }],
              },
            ],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render block and inline block nodes', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'block',
            item: 'block-123',
          },
          {
            type: 'paragraph',
            children: [
              { type: 'span', value: 'Text with ' },
              {
                type: 'inlineBlock',
                item: 'inline-456',
              },
              { type: 'span', value: ' inline block' },
            ],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should truncate long text content', () => {
      const longText = 'a'.repeat(100);
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            children: [{ type: 'span', value: longText }],
          },
          {
            type: 'code',
            code: longText,
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render paragraph with custom style', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            style: 'custom-style',
            children: [{ type: 'span', value: 'Styled paragraph' }],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    it('should render thematic break', () => {
      const root: Root = {
        type: 'root',
        children: [
          {
            type: 'paragraph',
            children: [{ type: 'span', value: 'Before break' }],
          },
          {
            type: 'thematicBreak',
          },
          {
            type: 'paragraph',
            children: [{ type: 'span', value: 'After break' }],
          },
        ],
      };

      const result = inspect(root);
      expect(result).toMatchSnapshot();
    });

    describe('with block formatter', () => {
      it('should format block with ID using custom formatter', () => {
        const root: Root = {
          type: 'root',
          children: [
            {
              type: 'block',
              item: 'block-123',
            },
          ],
        };

        const result = inspect(root, {
          blockFormatter: (item) => `ID: ${item}`,
        });
        expect(result).toMatchSnapshot();
      });

      it('should format block with object using custom formatter', () => {
        type BlockObject = {
          id: string;
          type: string;
          attributes: {
            title: string;
            content: string;
          };
        };

        const blockObject: BlockObject = {
          id: 'block-456',
          type: 'item',
          attributes: {
            title: 'My Block Title',
            content: 'Block content here',
          },
        };

        const root: Root<BlockObject> = {
          type: 'root',
          children: [
            {
              type: 'block',
              item: blockObject,
            },
          ],
        };

        const result = inspect(root, {
          blockFormatter: (item) => {
            if (typeof item === 'string') {
              return `ID: ${item}`;
            }
            return `${item.attributes.title} (${item.id})`;
          },
        });
        expect(result).toMatchSnapshot();
      });

      it('should handle multi-line block formatter output', () => {
        type ComplexBlockObject = {
          id: string;
          type: string;
          attributes: {
            title: string;
            description: string;
            metadata: { author: string; date: string };
          };
        };

        const blockObject: ComplexBlockObject = {
          id: 'block-789',
          type: 'item',
          attributes: {
            title: 'Complex Block',
            description: 'A detailed description',
            metadata: { author: 'John Doe', date: '2023-01-01' },
          },
        };

        const root: Root<ComplexBlockObject> = {
          type: 'root',
          children: [
            {
              type: 'paragraph',
              children: [{ type: 'span', value: 'Before block' }],
            },
            {
              type: 'block',
              item: blockObject,
            },
            {
              type: 'paragraph',
              children: [{ type: 'span', value: 'After block' }],
            },
          ],
        };

        const result = inspect(root, {
          blockFormatter: (item) => {
            if (typeof item === 'string') {
              return `ID: ${item}`;
            }
            return [
              `${item.attributes.title} (${item.id})`,
              `Description: ${item.attributes.description}`,
              `Author: ${item.attributes.metadata.author}`,
              `Date: ${item.attributes.metadata.date}`,
            ].join('\n');
          },
        });
        expect(result).toMatchSnapshot();
      });

      it('should format inline blocks with custom formatter', () => {
        type InlineBlockObject = {
          id: string;
          type: string;
          attributes: {
            name: string;
            text: string;
          };
        };

        const inlineBlockObject: InlineBlockObject = {
          id: 'inline-123',
          type: 'item',
          attributes: {
            name: 'Button',
            text: 'Click me',
          },
        };

        const root: Root<string, InlineBlockObject> = {
          type: 'root',
          children: [
            {
              type: 'paragraph',
              children: [
                { type: 'span', value: 'Click ' },
                {
                  type: 'inlineBlock',
                  item: inlineBlockObject,
                },
                { type: 'span', value: ' to continue' },
              ],
            },
          ],
        };

        const result = inspect(root, {
          blockFormatter: (item) => {
            if (typeof item === 'string') {
              return `ID: ${item}`;
            }
            return `${item.attributes.name}: "${item.attributes.text}"`;
          },
        });
        expect(result).toMatchSnapshot();
      });

      it('should fall back to default formatting when no formatter provided', () => {
        const root: Root = {
          type: 'root',
          children: [
            {
              type: 'block',
              item: 'block-123',
            },
            {
              type: 'paragraph',
              children: [
                {
                  type: 'inlineBlock',
                  item: 'inline-456',
                },
              ],
            },
          ],
        };

        // Test without formatter (should use default)
        const resultWithoutFormatter = inspect(root);
        expect(resultWithoutFormatter).toMatchSnapshot();

        // Test with empty options (should use default)
        const resultWithEmptyOptions = inspect(root, {});
        expect(resultWithEmptyOptions).toMatchSnapshot();
      });

      it('should provide max line width to formatter', () => {
        type BlockObject = {
          id: string;
          title: string;
          description: string;
        };

        const blockObject: BlockObject = {
          id: 'block-123',
          title: 'My Block',
          description:
            'This is a very long description that might exceed the suggested line width',
        };

        const root: Root<BlockObject> = {
          type: 'root',
          children: [
            {
              type: 'block',
              item: blockObject,
            },
          ],
        };

        const result = inspect(root, {
          blockFormatter: (item) => {
            if (typeof item === 'string') {
              return `ID: ${item}`;
            }

            return [
              item.title,
              `Description: ${item.description}`,
              `ID: ${item.id}`,
            ].join('\n');
          },
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
});
