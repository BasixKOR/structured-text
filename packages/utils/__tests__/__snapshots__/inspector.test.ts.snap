// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspector inspect should render a document wrapper 1`] = `
"└ paragraph
  └ span \\"Test\\"
"
`;

exports[`inspector inspect should render a simple root node 1`] = `
"└ paragraph
  └ span \\"Hello World\\"
"
`;

exports[`inspector inspect should render block and inline block nodes 1`] = `
"├ block (item: \\"block-123\\")
└ paragraph
  ├ span \\"Text with \\"
  ├ inlineBlock (item: \\"inline-456\\")
  └ span \\" inline block\\"
"
`;

exports[`inspector inspect should render code nodes with language and content 1`] = `
"├ code (language: \\"javascript\\") \\"console.log(\\"Hello, World!\\");\\"
└ code \\"echo \\"No language specified\\"\\"
"
`;

exports[`inspector inspect should render heading nodes with level 1`] = `
"├ heading (level: 1)
│ └ span \\"Main Title\\"
└ heading (level: 2)
  └ span \\"Subtitle\\"
"
`;

exports[`inspector inspect should render link nodes with URL and meta 1`] = `
"└ paragraph
  ├ span \\"Visit \\"
  ├ link (url: \\"https://example.com\\")
  │ └ span \\"this link\\"
  └ link (url: \\"https://meta.example.com\\", meta: {title=\\"Example Site\\", target=\\"_blank\\"})
    └ span \\"link with meta\\"
"
`;

exports[`inspector inspect should render list nodes with style 1`] = `
"└ list (style: bulleted)
  ├ listItem
  │ └ paragraph
  │   └ span \\"First item\\"
  └ listItem
    └ paragraph
      └ span \\"Second item\\"
"
`;

exports[`inspector inspect should render multiple children with correct tree structure 1`] = `
"├ paragraph
│ └ span \\"First paragraph\\"
└ paragraph
  └ span \\"Second paragraph\\"
"
`;

exports[`inspector inspect should render paragraph with custom style 1`] = `
"└ paragraph (style: \\"custom-style\\")
  └ span \\"Styled paragraph\\"
"
`;

exports[`inspector inspect should render span nodes with marks 1`] = `
"└ paragraph
  ├ span \\"Normal text\\"
  ├ span (marks: strong) \\"Bold text\\"
  └ span (marks: strong, emphasis) \\"Bold italic\\"
"
`;

exports[`inspector inspect should render thematic break 1`] = `
"├ paragraph
│ └ span \\"Before break\\"
├ thematicBreak
└ paragraph
  └ span \\"After break\\"
"
`;

exports[`inspector inspect should truncate long text content 1`] = `
"├ paragraph
│ └ span \\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...\\"
└ code \\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...\\"
"
`;

exports[`inspector inspect with block formatter should add TreeNode children to blocks without existing children 1`] = `
"└ block
  └ Block Info
    ├ Title: Block with TreeNode Children
    └ ID: block-789
"
`;

exports[`inspector inspect with block formatter should fall back to default formatting when no formatter provided 1`] = `
"├ block (item: \\"block-123\\")
└ paragraph
  └ inlineBlock (item: \\"inline-456\\")
"
`;

exports[`inspector inspect with block formatter should fall back to default formatting when no formatter provided 2`] = `
"├ block (item: \\"block-123\\")
└ paragraph
  └ inlineBlock (item: \\"inline-456\\")
"
`;

exports[`inspector inspect with block formatter should format block with ID using custom formatter 1`] = `
"└ block ID: block-123
"
`;

exports[`inspector inspect with block formatter should format block with object using custom formatter 1`] = `
"└ block My Block Title (block-456)
"
`;

exports[`inspector inspect with block formatter should format inline blocks with custom formatter 1`] = `
"└ paragraph
  ├ span \\"Click \\"
  ├ inlineBlock Button: \\"Click me\\"
  └ span \\" to continue\\"
"
`;

exports[`inspector inspect with block formatter should handle TreeNode returns for inline blocks 1`] = `
"└ paragraph
  ├ span \\"Click \\"
  ├ inlineBlock
  │ └ button (button-123)
  │   ├ Variant: primary
  │   └ Size: large
  └ span \\" to continue\\"
"
`;

exports[`inspector inspect with block formatter should handle blockFormatter returning a single TreeNode 1`] = `
"└ block
  └ Content: Block content here
    ├ Title: My Block Title
    └ ID: block-123
"
`;

exports[`inspector inspect with block formatter should handle blockFormatter returning an array of TreeNodes 1`] = `
"└ block
  ├ Properties
  │ ├ Title: Complex Block
  │ └ ID: block-456
  └ Metadata
    ├ Author: Jane Doe
    └ Tags
      ├ important
      └ feature
"
`;

exports[`inspector inspect with block formatter should handle multi-line block formatter output 1`] = `
"├ paragraph
│ └ span \\"Before block\\"
├ block Complex Block (block-789) Description: A detailed description Author: John Doe Date: 2023-01-01
└ paragraph
  └ span \\"After block\\"
"
`;

exports[`inspector inspect with block formatter should provide max line width to formatter 1`] = `
"└ block My Block Description: This is a very long description that might exceed the suggested line width ID: block-123
"
`;

exports[`inspector inspect with block formatter should treat string returns as single-line (strip newlines) 1`] = `
"└ block Title: Multi-line Block Description: Description with multiple lines ID: block-789
"
`;
